src/Language/Reflection/Types.idr

```diff
@@ -24,18 +24,30 @@ Res = Either String
 
 ||| Constructor of a data type
 public export
-record Con where
+record Con n where
   constructor MkCon
-  name : Name
-  args : List NamedArg
-  type : TTImp
+  name     : Name
+  args     : List NamedArg
+  typeArgs : Vect n TTImp
+
+||| True if the given constructor has only erased arguments.
+public export
+isConstant : Con n -> Bool
+isConstant = all isErased . args
+
+tcArgs : Elaboration m => Name -> (n : Nat) -> TTImp -> m (Vect n TTImp)
+tcArgs _  0     t@(IVar _ _) = pure []
+tcArgs nm (S k) (IApp _ s t) = (t ::) <$> tcArgs nm k s
+tcArgs nm _    _             = fail "Unexpected type for constructor \{nm}"
 
 ||| Tries to lookup a constructor by name.
 export
-getCon : Elaboration m => Name -> m Con
-getCon n = do (n',tt)    <- lookupName n
-              (args,tpe) <- unPiNamed tt
-              pure $ MkCon n' args tpe
+getCon : Elaboration m => (n : Nat) -> Name -> m (Con n)
```

Comment:
I think, function with the original semantics should be left, like

```idris
getAnyCon : Elaboration m => Name -> (n ** Con n)
```

because original function does not express from new one.

Reply:
Ah, I see, it's meant that type is already known prior the call to this function.
Shouldn't then `Con` be indexed by the type itself?

Reply:
> Ah, I see, it's meant that type is already known prior the call to this function. Shouldn't then `Con` be indexed by the type itself?

`Con` is indexed over the number of arguments of the corresponding type constructor the data constructor we inspect belongs to. This is to make sure all data constructors of a data type provide the same number of type-level arguments. I'm not sure what you mean by "indexed by the type itself"? Do you mean a `TTImp` index representing the data type the constructor belongs to? That wouldn't work, as indexed types define a family of types instead of a single one.

Reply:
I meant making type `Con` of kind `TypeInfo -> Type` instead of just `Nat -> Type`.

Reply:
> I meant making type `Con` of kind `TypeInfo -> Type` instead of just `Nat -> Type`.

Does this work? The list of constructors is a field in `TypeInfo`. I don't see the advantage of having such an index.

Reply:
Okay, now I understand that records cannot store such info even with mutual types, because we can't refer to _self_ from within a record.

Reply:
Anyway, what do you think about leaving a function with the original semantics, i.e. returning a constructor's info along with its params?

Reply:
For me, that's OK. I'll then just add an additional function for getting a constructor with the number of type arguments already known.